feg/gateway/services/envoy_controller/control_plane/control_plane.go

```diff
@@ -0,0 +1,342 @@
+	glog.Infof("Gateway listening HTTP/1.1 on port %d", port)
+	server := &http.Server{Addr: fmt.Sprintf(":%d", port), Handler: &xds.HTTPGateway{Server: srv}}
+	go func() {
+		if err := server.ListenAndServe(); err != nil {
+			glog.Error(err)
+		}
+	}()
+}
+
+func GetControllerClient() *ControllerClient {
+	cli := ControllerClient{}
+	ctx := context.Background()
+
+	glog.Infof("Starting Envoy control plane")
+
+	signal := make(chan struct{})
+	cb := &callbacks{
+		signal:   signal,
+		fetches:  0,
+		requests: 0,
+	}
+	cli.config = cache.NewSnapshotCache(mode == Ads, Hasher{}, nil)
+
+	srv := xds.NewServer(ctx, cli.config, cb)
+
+	// start the xDS server
+	go RunManagementServer(ctx, srv, port)
+	go RunManagementGateway(ctx, srv, gatewayPort)
+
+	<-signal
+
+	cb.Report()
+
+	return &cli
+}
+
+func (cli *ControllerClient) UpdateSnapshot(ues []*protos.AddUEHeaderEnrichmentRequest) {
+	nodeId := cli.config.GetStatusKeys()[0]
+
+	var clusterName = "cluster1"
+	cluster := []cache.Resource{
+		&v2.Cluster{
+			Name:                 clusterName,
+			ClusterDiscoveryType: &v2.Cluster_Type{Type: v2.Cluster_ORIGINAL_DST},
+			ConnectTimeout:       ptypes.DurationProto(6 * time.Second),
+			LbPolicy:             v2.Cluster_CLUSTER_PROVIDED,
+		},
+	}
+
+	filterChains := []*listener.FilterChain{}
+	var listenerName = "default_http"
+	var targetPrefix = "/"
+	var virtualHostName = "local_service"
+	var routeConfigName = "local_route"
+
+	glog.Errorf("Updating snapshot content")
+
+	for _, req := range ues {
+		var ue_ip_addr = string(req.UeIp.Address)
+		requestHeadersToAdd := []*core.HeaderValueOption{}
+
+		glog.Infof("Adding UE " + ue_ip_addr)
+
+		for _, header := range req.Headers {
+			requestHeadersToAdd = append(requestHeadersToAdd, &core.HeaderValueOption{
+				Header: &core.HeaderValue{
+					Key:   header.Name,
+					Value: header.Value,
+				},
+			})
+		}
+
+		virtualHosts := []*v2route.VirtualHost{}
+		virtualHosts = append(virtualHosts, &v2route.VirtualHost{
+			Name:                virtualHostName,
+			Domains:             req.Websites,
+			RequestHeadersToAdd: requestHeadersToAdd,
+			Routes: []*v2route.Route{{
+				Match: &v2route.RouteMatch{
+					PathSpecifier: &v2route.RouteMatch_Prefix{
+						Prefix: targetPrefix,
+					},
+				},
+				Action: &v2route.Route_Route{
+					Route: &v2route.RouteAction{
+						ClusterSpecifier: &v2route.RouteAction_Cluster{
+							Cluster: clusterName,
+						},
+					},
+				},
+			}}})
+
+		httpManager := &hcm.HttpConnectionManager{
+			CodecType:        hcm.HttpConnectionManager_AUTO,
+			StatPrefix:       "ingress_http",
+			UseRemoteAddress: &wrappers.BoolValue{Value: true},
+			CommonHttpProtocolOptions: &core.HttpProtocolOptions{
+				IdleTimeout: ptypes.DurationProto(3600 * time.Second),
+				//HeadersWithUnderscoresAction: core.HttpProtocolOptions_REJECT_REQUEST,
+			},
+			Http2ProtocolOptions: &core.Http2ProtocolOptions{
+				MaxConcurrentStreams:        &wrappers.UInt32Value{Value: 100},
```

Comment:
lets reduce this number to 16 or 32.
can you also define constant for this in beginning of this fine.